src/server/ns_turn_server.c

```diff
@@ -143,6 +146,93 @@ static int read_client_connection(turn_turnserver *server, ts_ur_super_session *
 
 static int need_stun_authentication(turn_turnserver *server, ts_ur_super_session *ss);
 
+
+/////////////////// rate limit //////////////////////////
+ur_addr_map *rate_limit_map = 0;
+TURN_MUTEX_DECLARE(rate_limit_main_mutex);
+
+void rate_limit_add_node(ioa_addr *address) {
+  // copy address
+  RateLimitEntry *rate_limit_entry = (RateLimitEntry*)malloc(sizeof(RateLimitEntry));
+  TURN_MUTEX_INIT(&(rate_limit_entry->mutex));
+  rate_limit_entry->request_count = 1;
+  rate_limit_entry->last_request_time = time(NULL);
+  rate_limit_entry->expiration_time = time(NULL) + RATE_LIMIT_ENTRY_EXPIRATION_SECS;
+
+  int request_count = 1;
+  int last_request_time = time(NULL);
+  int expiration_time = time(NULL) + RATE_LIMIT_ENTRY_EXPIRATION_SECS;
+
+  ur_addr_map_put_no_port(rate_limit_map, address, (ur_addr_map_value_type)rate_limit_entry);
+}
+
+ur_addr_map_cond_func ratelimit_delete_expired(ur_map_value_type value) {
+  time_t current_time = time(NULL);
+  RateLimitEntry *rate_limit_entry = (RateLimitEntry*)(void*)(ur_map_value_type)value;
+  if (rate_limit_entry->expiration_time < current_time)
+    return 1;
+  return 0;
+}
+
+int is_address_ratelimit(const ioa_addr *address) {
+  /* Housekeeping, prune the map when ADDR_MAP_SIZE is hit and delete expired items */
+  time_t current_time = time(NULL);
+
+  if (rate_limit_map == 0) {
```

Comment:
```suggestion
  if (rate_limit_map == NULL) {
```
While being technically the same, IMHO `NULL` expresses better that we're dealing with a pointer here.

Reply:
Fair point, NULL is best practices here.